controllers/idler/aap_idler.go

```diff
@@ -0,0 +1,252 @@
+	if err != nil {
+		return err
+	}
+	if aapGVR == nil {
+		// AAP API is not available/installed. Skipping.
+		return nil
+	}
+	aapAPI := aapAPI{
+		ResourceLists: resourceLists,
+		GVR:           *aapGVR,
+	}
+
+	// Check if there is any AAP CRs in the namespace
+	idledAAPs := []string{}
+	aapList, err := r.DynamicClient.Resource(aapAPI.GVR).Namespace(idler.Name).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if len(aapList.Items) == 0 {
+		// No AAP resources found. Nothing to idle.
+		return nil
+	}
+
+	// Get all pods running in the namespace
+	podList := &corev1.PodList{}
+	if err := r.AllNamespacesClient.List(ctx, podList, client.InNamespace(idler.Name)); err != nil {
+		return err
+	}
+	for _, pod := range podList.Items {
+		var aapName string
+		pod := pod // TODO We won't need it after upgrading to go 1.22: https://go.dev/blog/loopvar-preview
+		podLogger := log.FromContext(ctx).WithValues("pod_name", pod.Name, "pod_phase", pod.Status.Phase)
+		podCtx := log.IntoContext(ctx, podLogger)
+
+		// check the restart count for the pod
+		restartCount := getHighestRestartCount(pod.Status)
+		if restartCount > AAPRestartThreshold {
+			podLogger.Info("Pod is restarting too often for an AAP pod. Checking if it belongs to AAP and if so then idle the aap", "restart_count", restartCount)
+			aapName, err = r.ensureAAPIdled(podCtx, pod, aapAPI)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Check if running for longer then the AAP idler timeout
+			timeoutSeconds := r.aapTimeoutSeconds(idler)
+			if pod.Status.StartTime != nil && time.Now().After(pod.Status.StartTime.Add(time.Duration(timeoutSeconds)*time.Second)) {
+				podLogger.Info("Pod is running for too long for an AAP pod. Checking if it belongs to AAP and if so then idle the aap.", "start_time", pod.Status.StartTime.Format("2006-01-02T15:04:05Z"), "timeout_seconds", timeoutSeconds)
+				aapName, err = r.ensureAAPIdled(podCtx, pod, aapAPI)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// Check if we need to send a notification and proceed with the next pod
+		if aapName != "" {
+			// The AAP has been Idled
+			// A notification should be sent
+			if err := r.createNotification(podCtx, idler, aapName, "Ansible Automation Platform"); err != nil {
+				podLogger.Error(err, "failed to create Notification")
+				if err = r.setStatusIdlerNotificationCreationFailed(podCtx, idler, err.Error()); err != nil {
+					podLogger.Error(err, "failed to set status IdlerNotificationCreationFailed")
+				} // not returning error to continue tracking remaining pods
+			}
+
+			idledAAPs = append(idledAAPs, aapName)
+			if checkIfAllAAPsIdled(aapList, idledAAPs) {
+				// All AAPs are idled, no need to check the rest of the pods
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+func checkIfAllAAPsIdled(allAAPs *unstructured.UnstructuredList, idledAAPs []string) bool {
+	if len(allAAPs.Items) != len(idledAAPs) {
```

Comment:
I think the len will be different since we append the same aapName multiple times, see previous comment. We could maybe use a map for this, and this would also simplify the loops below.

Reply:
Great catch! Will fix it.

Reply:
Fixed in https://github.com/codeready-toolchain/member-operator/pull/631/commits/7c6e0e94ec34da50b737268382f9b3cc2c72b3a9